Feature/make radial menu great again

Fildrance opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Radial menu was great addition to UI but it requires precision as it is now. To bring up better UX it needs to be usable without too much precision so it can be used effectively in panic situations. This problem was already solved by wheel menu in most modern games - use sectors of circle to pick options and not icons themselves.

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Robust wheel UI!

## Technical details
<!-- Summary of code changes for easier review. -->

RadialMenu now uses 'special' button as context button in center - RadialMenuContextualCentralTextureButton. It considers space inside inner circle of parent RadialContainer half radius, and ALL outer space outside twice of its radius - as itself.
RadialContainer now puts angle sector data in each child  RadialMenuTextureButton.
RadialMenuTextureButton now can draw bagle segments!

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

[new-radia-menu.webm](https://github.com/user-attachments/assets/addb3dd8-45f8-45fe-8dce-770123c5adab)

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

removed RadialMenuButton - was not used.
now RadialMenuTextureButton requires to be inside RadialContainer to work properly

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl: Fildrance, SaphireLattice, eoineoineoin
- tweak: Radial menu UIs have been updated for readability.

Fildrance commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -93,13 +93,50 @@ protected override void Draw(DrawingHandleScreen handle)
+        // var internalCircle = new Vector2(Width /2f, Width / 2f);
+        // handle.DrawCircle(internalCircle, Radius/2, Color.Red, false);
+
+        base.Draw(handle);
```
was not calling base, dont start it now

Fildrance commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -93,13 +93,50 @@ protected override void Draw(DrawingHandleScreen handle)
+
+        base.Draw(handle);
+    }
+    public static float ConvertRadiansToDegrees(float radians)
```
remove unused code

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,12 +316,61 @@ private void OnClicked(ButtonEventArgs args)
+    /// <inheritdoc />
+    protected override bool HasPoint(Vector2 point)
+    {
+        var cX = -Position.X + Parent!.Width / 2;
```
cachable values

beck-thompson commented:
add a video to show everyone how cool this is

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -171,45 +179,37 @@ public void ReturnToPreviousLayer()
+    protected override bool HasPoint(Vector2 point)
     {
-        foreach (var ancestor in control.GetSelfAndLogicalAncestors())
+        if (Container == null)
```
to properly draw it considering radial menu button sectors - we need at least container itself (which contains radius data and can be different for each 'level' of menu).

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -171,45 +179,37 @@ public void ReturnToPreviousLayer()
+
+        var dist = Math.Sqrt(Math.Pow(point.X - cX, 2) + Math.Pow(point.Y - cY, 2));
+        var isNotInRadius = dist > Container.Radius * 2 || dist < Container.Radius / 2;
+        return isNotInRadius;
```
this DOES NOT respect not full circle menu! I feel like it will be real pain to make it work with with it. Maybe use default behaviour in this scenario and fallback to texturebutton mouse detection?

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -229,6 +252,69 @@ public RadialMenuTextureButton()
+
+        if (Parent is not RadialContainer container)
+        {
+            return;
```
following part is highly dependent on having radius from container so we will have to deal with not working at all if we are in invalid container.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -229,6 +252,69 @@ public RadialMenuTextureButton()
+        var segCount = (int)(controlSegmentSize / singleSegmentSize) + 1;
+
+        // CHANGE TO STACKALLOC AND MOVE THIS STUFF TO TOOLBOX!111
+        var buffer = new Vector2[segCount * 2];
```
i would like to move it to DrawingHandleScreen but it feels like i'll have to write implementation for Clide renderhandler too? I really dont want to do that, it looks kinda more complex then what i have done here. But using stackalloc is kinda required here imo :(

Fildrance replied:
Sloth said move it to DrawingHandleScreen

Fildrance replied:
PJB said its too specific, so i'm leaving it here. And buffer is now pre-allocated most of the time.

Fildrance commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,36 +67,86 @@ public RadialContainer()
+                    Radius * 2 * separatorSin + halfWidth,
+                    -Radius * 2 * separatorCos + halfHeight
+                ) * UIScale,
+                new Color(173, 216, 230, 180) // todo: use stylesheets
```
this is not button but i'd like to use styles from buttons! Buttom background color when its hovered, to be precise! I don't really understand how i can do it - TryGetStyleProperty is working in a wierd way even when used in proper class hierarchy, and here it is quite empty Q_Q

chromiumboy replied:
The colors can be defined as separate fields instead of styles

Fildrance replied:
Yea, this PR is getting bloated, we will have to just leave it as fields with default values, i'm not going for stylesheets. Sadge.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -229,6 +252,69 @@ public RadialMenuTextureButton()
+            buffer[i * 2 + 1] = containerCenter + point * radius / 2;
+        }
+
+        var color = DrawMode == DrawModeEnum.Hover
```
same as stated before about stylesheets but with 1 more problem - this is TextureButton but i need styles for ContainerButton coz i think only it have background colors that i can use. Should i create separate bunch of code in stylesheets just for that? It feels wierd :(

chromiumboy replied:
If you want them, then yes

eoineoineoin replied:
Maybe too lofty for this PR, but it sure would be nice to be able to style the segments using something similar to a `StyleBox` - I'd really like to see a `StyleBoxTexture` used to apply more than just a solid colour to the segment.

Fildrance replied:
i'll leave Color properties for now, i think i spent way too much with this PR - styling in the way i was hoping for is not possible due to sytlesheet mechanism limitations, and adding new style elements to all themes is not what i want in this pr - it is already starting to bloat! We need to merge it before it is dead.

Fildrance commented on Content.Client/Chat/UI/EmotesMenu.xaml:
```diff
@@ -1,4 +1,4 @@
-﻿<ui:RadialMenu xmlns="https://spacestation14.io"
+<ui:RadialMenu xmlns="https://spacestation14.io"
```
what the hell changed? i cannot see even whitespace changes! WTF

0x6273 replied:
UTF8 bom moment, don't worry about it

Fildrance commented:
THIS DEPENDS ON https://github.com/space-wizards/RobustToolbox/pull/5479

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -229,6 +252,69 @@ public RadialMenuTextureButton()
+    {
+        var texture = TextureNormal;
+
+        if (texture == null)
```
maybe draw colored segment even if no texture was passed?

chromiumboy replied:
Yeah, that's fine. I think it actually looks better without the button texture

Fildrance replied:
now segment drawing is field that can be changed per-button (make zebra control for fun?) and button can be rendered with background, with only border etc.

chromiumboy replied:
Just drop the button texture entirely, it clashes with the sector background

Fildrance replied:
Well now i understand what u meant. Honestly i was not realizing it was problem until it was gone :D

chromiumboy commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
The base radial container shouldn't include dividers, as someone might want to embed it into a control for purposes other than radial menus. It was designated as virtual for that reason; I'd create a new radial container class that inherits from this and override the draw method

Fildrance replied:
moved border drawing into segmented buttons - its just more fitting out there.

Fildrance replied:
created separate type for segmented button in radial menu - they can both exist in radial menu now!

chromiumboy commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,36 +67,86 @@ public RadialContainer()
 		
 		// Add padding from the center at higher child counts so they don't overlap.
-		Radius = baseRadius + (childCount * radiusIncrement);
+		Radius = baseRadius + (elementCount * radiusIncrement);
```
This adjustment changes the overall area of the container, so there's a large disparity in size between menu layers that contain only a few elements and those that contain many. Consider having a set size for the sectors and only shifting the positions of the icons

On the topic of size, the amount of screen real estate that the sectors take up is quite large, consider reducing the outer radius as well, if it doesn't cause issues with the above

Fildrance replied:
i'll remove separators from radius increment - that makes sense, it was my failure to properly understand why it is there.
About size of sectors and 'free-space' - the point of this UX changes was to make it require less precision - so you can click it faster and not really aim, but make choice of direction. Size is just for that - most of games have this 'wheel' ui blur and grow in opacity from center to outer corners - i could not make it work here, sadly (tbf was not trying) but REAL size of radial menus were quite the same overall (if we remove blur and make controls constant color), i belive!
![image](https://github.com/user-attachments/assets/329cae48-6066-4fec-9c07-e47a796998bd)

Fildrance replied:
If there is some proposal about how to change numbers exactly - you are welcome to ask and i will try it - currently its radius * 2 for outer, radius /2 for inner

Fildrance replied:
Changed back to only use child count thanks to eoin. Kinda missed it initially. Yay!.
Multipliers were moved to properties - so can be easily changed.

ps3moira commented:
is there a way to make it smaller? This looks way too big

gusxyz commented:
> is there a way to make it smaller? This looks way too big

Agreed theres a lot of deadspace.

eoineoineoin commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -172,55 +180,66 @@ public void ReturnToPreviousLayer()
 
-        return null;
+        var cX = -Position.X + Parent!.Width / 2;
+        var cY = -Position.Y + Parent.Width / 2;
```
This is almost certainly supposed to be Parent.Height?

Fildrance replied:
yeap, fixed that

eoineoineoin commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
+    protected override bool HasPoint(Vector2 point)
+    {
+        var cX = -Position.X + Parent!.Width / 2;
+        var cY = -Position.Y + Parent.Width / 2;
```
This is almost certainly supposed to be Parent.Height?

eoineoineoin commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
+        var bufferSize = segmentCount * 2;
+        if (_sectorPointsForDrawing == null || _sectorPointsForDrawing.Length != bufferSize)
+        {
+            _sectorPointsForDrawing ??= new Vector2[bufferSize];
```
This looks like it's trying to cache the calculated points, but ends up calculating them every time anyway?

Fildrance replied:
it is pre-allocating array once and not on every frame - so we lessen GC pressure

Errant-4 reviewed (commented):
There appears to be a narrow "gap" everywhere the segments join, where it is possible to click through the interface and act on the entity underneath

Fildrance commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,36 +67,86 @@ public RadialContainer()
+            var childY = -Radius * MathF.Cos(targetAngleOfChild) + halfHeight - child.Height / 2f;
+            var position = new Vector2(childX, childY);
+
             SetPosition(child, position);
```
This stuff is not even getting recalculated every frame! Its magic!

Fildrance commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,36 +67,86 @@ public RadialContainer()
             SetPosition(child, position);
+
+            // radial menu buttons need to know in which sector around container they should be rendered
+            if (child is RadialMenuTextureButton tb)
```
the best idea i can imagine to counter that - is to creat interface. It have 95% chance to not ever be used, but architectural remark is correct. So it will be done.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
+        var cX = -Position.X + Parent!.Width / 2;
+        var cY = -Position.Y + Parent.Width / 2;
+
+        var dist = Math.Sqrt(Math.Pow(point.X - cX, 2) + Math.Pow(point.Y - cY, 2));
```
Thank, looks great and so much easier to understand, thanks

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
+
+        var dist = Math.Sqrt(Math.Pow(point.X - cX, 2) + Math.Pow(point.Y - cY, 2));
+        var parent = (RadialContainer)Parent;
+        var isInRadius = dist < parent.Radius * 2 && dist > parent.Radius / 2;
```
fixed

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -229,6 +248,42 @@ public RadialMenuTextureButton()
+
+        // draw sector where space that button occupies actually is
+        var pX = -Position.X + Parent!.Width / 2;
+        var pY = -Position.Y + Parent.Width / 2;
```
fixed

Fildrance commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,36 +67,86 @@ public RadialContainer()
+            handle.DrawLine(
+                new Vector2(
+                    Radius/2 * separatorSin + halfWidth,
+                    -Radius/2 * separatorCos + halfHeight
```
This is a problem with UI coordinate matrics kinda 'flipped' compared to how normally you imagine coordinates - X is from left to rigth, but Y is from top to bottom, so +∞ / +∞ is bottom-right, which is making all this trickery. I honestly not sure that its worth effert to flip coordinates just for those calculations. Also i am not sure how to do the flip and not break everything / add basically useless caculations on every frame.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -172,55 +180,66 @@ public void ReturnToPreviousLayer()
+        var cX = -Position.X + Parent!.Width / 2;
+        var cY = -Position.Y + Parent.Width / 2;
+
+        var dist = Math.Sqrt(Math.Pow(point.X - cX, 2) + Math.Pow(point.Y - cY, 2));
```
good point, ty for reminding :3

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
+
+        var dX = cX - point.X;
+        var dY = cY - point.Y;
+        var angle = dX > 0
+            ? -MathF.Atan2(dX, dY) + MathF.PI * 2
+            : -MathF.Atan2(dX, dY);
```
as i said in other comment - its all due to Y axis being flipped - i really dont want to dig it, i'm not used enough to working with graphics to efficiently cope with such problems >_> And it does not seem like its that critical - radial controls will improve MUCH more just from automating working of UI layer with model - just look at EmotesMenu c-tor, its bloated shit, 70% of code is stuff that should be solved by infrastructure and not high-level control itself...

eoineoineoin replied:
By checking in bad math, you're creating _two_ problems for the next person to work on this code, where now, they need to understand your code and then they need to fix it, before they can even get started on whatever they were trying to do.

I have made your math more conventional in https://github.com/eoineoineoin/space-station-14/commit/7ec22897a66c997cca42e433320e94429808cc15, if you want to cherry-pick that.

Fildrance replied:
There is an interesting side effect of this change set - it looks like sections rotated by a 30 degrees. I'm quite stunned how it was changed but it still works >_> need to look into math more...
![image](https://github.com/user-attachments/assets/6cac1929-2177-428f-a451-e9be7f414349)

Fildrance replied:
Okay its actually mirrored (that part i understand) and then rotated 90 degrees. I'm not sure how critical of a change that would be - people most likely got accustomed to current icon positions...

Fildrance replied:
I had to make some corrections - as chromiumboy said in discord - its better to have it oriented around Y-axis symmetry, not X-asix symmetry. So i added offset as button parameter. Checked - it works both ways, rotates stuff. yay.
And separator lanes logic was messed up with offset because of this offset. Oh well, now angle is negative here and all works well. I'd say that i kept most of your changes :3 so ty for contribution to our great cause.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -172,55 +180,66 @@ public void ReturnToPreviousLayer()
 
-        return null;
+        var cX = -Position.X + Parent!.Width / 2;
+        var cY = -Position.Y + Parent.Width / 2;
```
tbh i have low hopes that injecting stuff like parent center from other control and not getting it here will improve at least something. Controls ARE very coupled and it is not really bad thing in itself. Bad thing is inability to easily control nesting from compiler standpoint - those controls are not viable outside of radial container - THEY SHOULD NOT BE, but as we cannot mark it really - well, at least with not calling 'Parent' it is prettier.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -229,6 +248,42 @@ public RadialMenuTextureButton()
+
+        // draw sector where space that button occupies actually is
+        var pX = -Position.X + Parent!.Width / 2;
+        var pY = -Position.Y + Parent.Width / 2;
```
done

Fildrance commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,36 +67,86 @@ public RadialContainer()
         {
-            var position = new Vector2(Radius * MathF.Sin(AngularRange.X + sepAngle * i) + Width / 2f - child.Width / 2f, -Radius * MathF.Cos(AngularRange.X + sepAngle * i) + Height / 2f - child.Height / 2f);
+            // odd indexes are child elements (buttons)
+            var indexForChildElement = index * 2 + 1;
```
Sorry i currently cannot wrap my head around on how code should be to get rid of using separate index for overall segments and child elements. Can you explain you idea?

eoineoineoin replied:
I have simplified the code in https://github.com/eoineoineoin/space-station-14/commit/203953e4e170e0673526a0f6c5bf15e16f5627e9, if you want to cherry-pick that into your branch.

Fildrance replied:
Ok i have to say that using childCount*0.5 was an option i was not seeing in plain sight. Honestly difference in understadnability is not that grat but it exists, no reason to not use that code, ty.

Fildrance commented:
> is there a way to make it smaller? This looks way too big

Kinda made it smaller (by reducing). But the whole point is to make control which is NOT requiring precise clicking, so there is merit about how much smaller it can be and still retain this UX feature. Currently it is radius of original control (circle on which icons resides) devided by 2 (inner border) and multiplied by 2 (outer border).

Fildrance commented:
> There appears to be a narrow "gap" everywhere the segments join, where it is possible to click through the interface and act on the entity underneath
> 
>  gap.mp4

Now it could be fixed. Potentially.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
+        for (var i = 0; i < segmentCount; i++)
+        {
+            float angle;
+            if (i == segmentCount - 1)
+            {
+                // fix rounding problem that was created when calculating count of segments as int
+                angle = angleSectorTo;
+            }
+            else
+            {
+                angle = angleSectorFrom + minimalSegmentSize * i;
+            }
```
No, we are dealing with float rounding, this is not gonna work this way i am mostly sure.

Fildrance replied:
Actually, i just floor-ed values when setting position of control from radial container - and that fixed most of the stuff. So your solution works nicely now. wierd.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -172,55 +180,66 @@ public void ReturnToPreviousLayer()
+        // Button space is inside half of container radius / or outside double of its radius.
+        // half of radius and double the radius are radial menu concentric circles that are
+        // created by radial menu buttons.
+        return dist > ActiveContainer.Radius * 2 || dist < ActiveContainer.Radius / 2;
```
done

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
+    /// <param name="angleSectorTo">Angle in radian, from which sector should start.</param>
+    /// <param name="color">Color for drawing.</param>
+    /// <param name="filled">Should figure be filled, or have only border.</param>
+    private void DrawBagleSector(
```
fixed it as bagle is kinda 3d object and it was misleading

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -242,14 +297,107 @@ private void OnClicked(ButtonEventArgs args)
+                angle = angleSectorFrom + minimalSegmentSize * i;
+            }
+
+            var point = new Angle(angle).RotateVec(-Vector2.UnitY);
```
I can leave todo for optimization. Will this do? I feel using transfrom can bring up pretty misearable bugs. Other then that i could not figure out how to use setTransform based on overloads it provides and data i have. 
I also must remark that it currently calculates almost constant count of points ~32 for whole circle, and it is not really depending on amount of segments circle should be split up on.

Fildrance replied:
Most likely i'll look at using shaders, sloth said it'll be better. But not in this PR, it suffered enough! :BibleThump:

metalgearsloth reviewed (changes requested):
Isn't it better to just use a shader to avoid the segment overlap? Will also be much cheaper to draw.

Fildrance commented:
> Isn't it better to just use a shader to avoid the segment overlap? Will also be much cheaper to draw.

Can this be idea for future improvement? In separate PR. 
I have zero idea how to work with shaders and i even was thinking that shaders are not something that CAN work for such scenario. Resources that drawring requires for now looks fine, overlapping of segments that Errant was talking about is not drawing problem and it should not exist currently.

chromiumboy commented:
> Isn't it better to just use a shader to avoid the segment overlap? Will also be much cheaper to draw.

Most of our contributers won't have experience with writing shaders, so I don't we can really request that they do so

lzk228 commented:
can you also move text from on hover to make him permanent under or below icons?

Fildrance commented:
> can you also move text from on hover to make him permanent under or below icons?

I belive that part of hover behaviour is from parent component (TextureButton) so that would need... some copy-paste from there. Also not all tooltips would look nice, i feel >_> which is out of current PR scope. I'd better have separate issue for that thing to be fixed and reviewed without all other garbage i coded here

Fildrance commented:
@eoineoineoin can u please resolve stuff i that was fixed and where answer statisfied you?

eoineoineoin commented:
> @eoineoineoin can u please resolve stuff i that was fixed and where answer statisfied you?

I don't have any ability to resolve those threads, I'm afraid.

Did you see these comments?
https://github.com/space-wizards/space-station-14/pull/32653#discussion_r1798160753
https://github.com/space-wizards/space-station-14/pull/32653#discussion_r1798158503

Fildrance commented:
> > @eoineoineoin can u please resolve stuff i that was fixed and where answer statisfied you?
> 
> I don't have any ability to resolve those threads, I'm afraid.
> 
> Did you see these comments? [#32653 (comment)](https://github.com/space-wizards/space-station-14/pull/32653#discussion_r1798160753) [#32653 (comment)](https://github.com/space-wizards/space-station-14/pull/32653#discussion_r1798158503)

Thank you! I will look at that!

Fildrance commented:
> There appears to be a narrow "gap" everywhere the segments join, where it is possible to click through the interface and act on the entity underneath
> 
>  gap.mp4

Ok i found out why it was happening - float was messing up numbers. I floored position on menu item controls - it snapped to physical grid and al of the math works perfecly now - no holes, not in graphycs, not in HasPoint checks!

Fildrance commented:
> Any chance you can have the option to chance opacity on it? That looks too obnoxious for me to use.

Color (and as result opacity) is a public property of control, so it can be changed. What exact values to use by default - dunno ;(

chromiumboy commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -205,51 +278,308 @@ private void OnClicked(ButtonEventArgs args)
+    /// <summary>
+    /// Color of background in non-hovered state.
+    /// </summary>
+    public Color BackgroundColor { get; set; } = new Color(173, 216, 230, 70);
```
```suggestion
    public Color BackgroundColor { get; set; } = new Color(70, 73, 102, 128);
```

chromiumboy replied:
Leaving some suggestions to hopefully better match the exist UI color palette

Fildrance replied:
done.

chromiumboy commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -205,51 +278,308 @@ private void OnClicked(ButtonEventArgs args)
+    /// <summary>
+    /// Color of background in hovered state.
+    /// </summary>
+    public Color HoverBackgroundColor { get; set; } = new Color(173, 216, 230, 100);
```
```suggestion
    public Color HoverBackgroundColor { get; set; } = new Color(87, 91, 127, 128);
```

chromiumboy commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -205,51 +278,308 @@ private void OnClicked(ButtonEventArgs args)
+    /// Color of separator lines.
+    /// Separator lines are used to visually separate sector of radial menu items.
+    /// </summary>
+    public Color SeparatorColor { get; set; } = new Color(173, 216, 230, 180);
```
```suggestion
    public Color SeparatorColor { get; set; } = new Color(128, 128, 128, 128);
```

Fildrance replied:
done.

chromiumboy commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -205,51 +278,308 @@ private void OnClicked(ButtonEventArgs args)
+        bool filled = true
+    )
+    {
+        const float minimalSegmentSize = MathF.Tau / 32;
```
```suggestion
        const float minimalSegmentSize = MathF.Tau / 128f;
```

You can afford some extra triangles

Fildrance replied:
Well if you say so. It was looking KINDA ugly. done

chromiumboy commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,37 +67,77 @@ public RadialContainer()
-		Radius = baseRadius + (childCount * radiusIncrement);
+
+        // Add padding from the center at higher child counts so they don't overlap.
+        Radius = baseRadius + (childCount * radiusIncrement);
```
You still have the issue with the sector radii being dependent on the number of children in the container

Also, this is going to override anyone setting the radius of the container in XML code

It should be 'AdjustedRadius = Radius + (childCount * radiusIncrement);' or similar

Fildrance replied:
@chromiumboy yes but that was in original control design - i have no idea why that was made. Are you saying that it should not be here now and this was a mistake? Or that for this type of buttons it should not be used?

chromiumboy replied:
Radius being overwritten sounds like it was an oversight by the original coder. It needs to be corrected

Fildrance replied:
Oh it wasn't you, i'm sorry! Yea, i'll change it so it'll use provided from property radius as BaseRadius. Also i'll need to adjust radius that most controls were using - they used smaller numbers as base valus. Otherwise they will dramatically change in size - and we don't want such sudden change to stumble users...

chromiumboy commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,37 +67,77 @@ public RadialContainer()
+    protected override Vector2 ArrangeOverride(Vector2 finalSize)
     {
-		
         const float baseRadius = 100f;
```
Constants should be defined at the top of the file

Fildrance replied:
need to understand what to do with logic that adds padding by incrementing based on child count - maybe baseRadius is a bad idea? or it should be MinRadius?

Fildrance replied:
Renamed Radius with InitialRadius, and old Radius property is CalculatedRadius now, xml-doc explains that it depends on InitialRadius and uses visible child count and const RadiusIncrement to be calculated. CalculatedRadius have private setter for safety reasons

chromiumboy commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,37 +67,77 @@ public RadialContainer()
     {
-		
         const float baseRadius = 100f;
         const float radiusIncrement = 5f;
```
See above

Fildrance replied:
Extracted const.

Fildrance commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,37 +67,77 @@ public RadialContainer()
+                tb.AngleSectorTo = sepAngle * (childIndex + 1);
+                tb.AngleOffset = angleOffset;
+                tb.InnerRadius = Radius / 2;
+                tb.OuterRadius = Radius * 2;
```
extracting property for multipliers - thats actually good idea : > 
about 1.5f - ok, but bear in mind WHY we are adding those changes - we want control to be not requiring precision, so you can use it very-very fast. Making it too small make it not UX but just visual change. But having it variable - we can change it in any moment - so its fine.

Fildrance commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -205,51 +278,308 @@ private void OnClicked(ButtonEventArgs args)
+        var type = filled
+            ? DrawPrimitiveTopology.TriangleStrip
+            : DrawPrimitiveTopology.LineStrip;
+        drawingHandleScreen.DrawPrimitives(type, _sectorPointsForDrawing, color);
```
having to call it on each draw is a bit wierd. But ok. fixed.

chromiumboy replied:
If it's too expensive here, move it else where,or use a lookup table to convert between rgb and srgb colors

Fildrance replied:
Its not like 'too expensive', but we shove everything that sure we can be cached to the cache for perf. Not nice thing is that it will require some wierd getter-setter dancing. I don't mind it, gonna do it...

Fildrance replied:
done

Fildrance commented on Content.Client/RCD/RCDMenu.xaml:
```diff
@@ -12,21 +12,21 @@
 
     <!-- Entry layer (shows main categories) -->
     <ui:RadialContainer Name="Main" VerticalExpand="True" HorizontalExpand="True" Radius="64" ReserveSpaceForHiddenChildren="False">
-        <ui:RadialMenuTextureButton StyleClasses="RadialMenuButton" SetSize="64 64" ToolTip="{Loc 'rcd-component-walls-and-flooring'}" TargetLayer="WallsAndFlooring" Visible="False">
```
ah, now i see what u meant - yea, i removed from all of the stuff that is currently used but style code will remain in case someone would like to use control simple texture buttons that i left as an option :>

chromiumboy commented on Content.Client/UserInterface/Controls/RadialMenu.cs:
```diff
@@ -205,51 +278,329 @@ private void OnClicked(ButtonEventArgs args)
+    /// <summary>
+    /// Color of background in non-hovered state. Accepts RGB color, works with sRGB for DrawPrimitive internally.
+    /// </summary>
+    public Color BackgroundColor
```
Hmm, this is a bit sussy. It's probably better if we just call this field 'BackgroundColorSrgb' (same for those below) and expect people to supply a srgb formatted value rather than this conversion back and forth. Sorry 'bout misleading you on that

Fildrance replied:
There is not 'srgb c-tor', so its just plain harder to supply correct  thing - i see no reason to enfoce developer to do more action that is actually required. Stylesheet code includes only RGB and not sRGB colors as far as i can see at first glance, so if it will be applied - it should break notion just for this control, and also contain more code then it is crucial to declare stylesheet info (transform to srgb). Getter/Setter logic IS bad when you are going to apply it every second /tick or so, but thats clearly not he case, right?

chromiumboy replied:
OK. Let's leave it for now and see what any other rmaintainer thinks

Fildrance commented:
enabled any functional keys pressed when pushing radial menu buttons as AI requires alt to be pressed when opening door menu - this will make it feel more fluent as now u dont need release alt when clicking menu, and then press it again to call menu on next door.

Fildrance commented:
Updated video because i corrected sector size and colors as chrom asked some time ago.

Pumkin69 commented:
this should be smaller and more transparent imo it covers too much

Errant-4 commented:
> this should be smaller and more transparent imo it covers too much

Why is that a problem for an interface that only exist for the duration and purpose of the player selecting something on it?

Fildrance commented:
> Why is that a problem for an interface that only exist for the duration and purpose of the player selecting something on it?

I'd say even more - the purpose of this is to be able to select something without precision, as was said MULTIPLE times! So the larger segments are - less precision you need. It was designed to be able to click right section while u are running in merely 0,3-0,5 seconds! This neat detail can introduce more skill play in some places, where we will see it fitting (did sm1 say minigames?), and overall will update UX.

This menu however is not supposed to stay on you ui for long, and for this reason any click outside of menu will be considered as back/cancel action. So if it is not needed currently, u can in the same manner (very fast) click in center of menu, punch ESC or click in any point outside of window, without picking menu any options. I think its very neat as it is, and it is very small right now. But oh well.

Errant-4 reviewed (dismissed):
It seems that when you increase UI scale the icons get larger, but the Disc does not 

![Screenshot_2024-12-02_102457](https://github.com/user-attachments/assets/c457c9c4-5427-4df7-8894-a62a3398dee3)
![Screenshot_2024-12-02_102425](https://github.com/user-attachments/assets/a2402954-ec78-4932-806a-db7e15cfbe8c)

Fildrance commented:
> It seems that when you increase UI scale the icons get larger, but the Disc does not

Ty for your input Errant, i found and fixed it - forgot to use UIScale when rendering sectors of menu

Errant-4 commented:
Also, clicking the Back button on a submenu exits the UI rather than going back to the previous level

Fildrance commented:
> Also, clicking the Back button on a submenu exits the UI rather than going back to the previous level

Really neat bug! I wasted some time trying to understand problem and finally i got it - i use _enableAllKeybinds flag so clicking with 'alt' key works too (for AI menu its really important, i tested it) but it enables propagation for more of 'click' events then just 'click' as a result. And as click also invokes MoveStoredItem - so 'back button' is called twice, and second call, as intended, closes menu.

SlamBamActionman reviewed (changes requested):
Mmm we love ourselves some nice UI 👍 

1. As you've noticed in the comment above, the UI is currently triggering on more inputs than desired. While testing it basically fired on any keypress that also activated something else (walking, activating the chat window etc). This must be fixed before it can be merged.

2. My second point of feedback is probably more of a design opinion. With the size of the radial menu being quite permissive and clear now, I'd probably prefer if clicking outside the circle closes the menu completely, while clicking on the middle circle goes back one step. When you're in the first level of the menu this behavior ends up being the same and feels nice as-is, but it did feel odd trying to close it when you're in the second level and needing to click twice. This leaves the Escape key as the only option to instantly close it. It also looked a bit odd that the middle button ends up being highlighted when the mouse is far outside the circle. Since we already have two different areas (inside the circle vs. outside the circle), why not give those different functionalities?

Fildrance commented:
```
As you've noticed in the comment above, the UI is currently triggering on more inputs than desired. While testing it basically fired on any keypress that also activated something else (walking, activating the chat window etc). This must be fixed before it can be merged.
```
I thought i fixed it already - see commit just before your comment. Can u say how to repro problem? >_>

```
My second point of feedback is probably more of a design opinion. With the size of the radial menu being quite permissive and clear now, I'd probably prefer if clicking outside the circle closes the menu completely, while clicking on the middle circle goes back one step. When you're in the first level of the menu this behavior ends up being the same and feels nice as-is, but it did feel odd trying to close it when you're in the second level and needing to click twice. This leaves the Escape key as the only option to instantly close it. It also looked a bit odd that the middle button ends up being highlighted when the mouse is far outside the circle. Since we already have two different areas (inside the circle vs. outside the circle), why not give those different functionalities?
```
Well, if we will make it react to outside clicks as separate function, then there would be a problem - i'd need to either:
1. Create a separate control, which will be hidden button and will cover all area outside of radial menu outer radius. Hidden controls like that is a workable solution, but it really is not pretty
2. Make central button BEHAVE differently when click outside is done - sure, works, but highlights button when u have mouse outside STILL.
3. Override click code for radial menu control itself - well, this is going to be a very-very dirty trip, as click is composition of 2 events - mouse down and up, and some tracking (checking if up was at the same 'area'/control as down was previously). 

not a fan of any of these tbh... are we sure we want this? i don't hate the idea, i'd just say that its not really worth it...

VerinSenpai commented:
This is a nice qol change in my opinion. I have no issue with the size as it is but perhaps something in settings like radial menu size to make it adjustable to peoples liking.

SlamBamActionman commented:
> I thought i fixed it already - see commit just before your comment. Can u say how to repro problem? >_>

Not sure what steps I can give. It happens whenever I hover my mouse over an option in the circle (does not happen inside/outside the circle) and it acts as if I clicked on it if I provide any other input.

1. I open up the emote menu with Y
2. I move my cursor over an option, like "Hands"
3. I press W, which makes me walk upwards and also select "Hands" at the same time.

> Well, if we will make it react to outside clicks as separate function, then there would be a problem - i'd need to either:
> 
> 1. Create a separate control, which will be hidden button and will cover all area outside of radial menu outer radius. Hidden controls like that is a workable solution, but it really is not pretty
> 2. Make central button BEHAVE differently when click outside is done - sure, works, but highlights button when u have mouse outside STILL.
> 3. Override click code for radial menu control itself - well, this is going to be a very-very dirty trip, as click is composition of 2 events - mouse down and up, and some tracking (checking if up was at the same 'area'/control as down was previously).
> 
> not a fan of any of these tbh... are we sure we want this? i don't hate the idea, i'd just say that its not really worth it...

The way I would do it is make the central circle its own clickable area. 
![image](https://github.com/user-attachments/assets/8792f8b1-91fc-4124-b2cc-56025bf742aa)
Imagine that the red circle is its own button, and that is the area you can click to go back in the UI (don't put an actual red circle there, keep it looking them same). Clicking anywhere else (which would then have to be outside the whole circle) will close the UI fully. That should be doable, right?

Fildrance commented:
Fixed controls handling any input when mouse is hovering them (now only click or alt-click).
Added new button for 'outside' area. New button is iffy, but i'm not sure if alternatives are worse... will be happy for some advice on that...

SlamBamActionman reviewed (changes requested):
I've tested it now and I'm liking the interaction flow - just some small things that should be fixed:

1) For some reason, holding the RCD and opening up the menu using the item use key (Z for me) only keeps the menu open if you keep holding the button; releasing it closes the menu with no selection. This is different from current behavior, where pressing the button keeps the menu open and pressing it again closes it. Opening/closing the emote meny (using Y), or clicking on the RCD in your hand, correctly does the toggle behavior though.

2) You can get multiple radial menus open, which looks a bit weird. They still behave as expected though, with the latest opened menu being dealt with first. Ideally you'd probably want a new menu close the old one, but if that's difficult to get in I think this may still pass. 
![image](https://github.com/user-attachments/assets/ec57b1a0-3670-43c4-81c4-3af8e09f682c)

Fildrance commented:
Could not reproduce problem with menu requiring HOLDING button to stay open - this might be caused by not upding local branch - pls check if previous bug (when any button press, inluding move, was considered as click on radial).

Errant-4 commented:
I can't unsee that the outer diameter of the rings is different depending on how many buttons there are on it

metalgearsloth commented on Content.Client/UserInterface/Controls/RadialContainer.cs:
```diff
@@ -67,37 +91,74 @@ public RadialContainer()
+        var controlCenter = finalSize * 0.5f;
 
         // Adjust the positions of all the child elements
-        foreach (var (i, child) in children.Select((x, i) => (i, x)))
+        var query = children.Select((x, index) => (index, x));
+        foreach (var (childIndex, child) in query)
         {
```
Am I missing something or is this select pointless?

Fildrance replied:
it binds item with its index? but yes! you are correct, removing linq here won't hurt...

Fildrance replied:
Ah, no, children is not ICollection of any kind, its IEnumerable due to having filter for visibility in line 100.
So i cannot replace it with for cycle Q_Q

Fildrance commented:
> I can't unsee that the outer diameter of the rings is different depending on how many buttons there are on it 
That actually was done before me, i'm just displaying the same logic, trying not to change it. There is SOME degree of reason there - the more items there is, the more distance between them u need for icons to simply not collide! But all numbers (radius increment per item) are a bit random on that part, if you ask me. It also undermines a bit existance of 'radius' property on control. BUT! Its helping in case u set radius on control but have very different amount of items in different SUB-menus. Kinda.

Kirill2253 commented:
Unity Station🥶
